airbase/download_api.py

```diff
@@ -0,0 +1,319 @@
+
+
+async def _raise_for_status(response: httpx.Response):
+    response.raise_for_status()
+
+
+async def _warn_for_status(response: httpx.Response):
+    if response.status_code >= 400:
+        warn(
+            f"HTTP Error {response.status_code} for URL {response.url}",
+            category=RuntimeWarning,
+        )
+
+
+class DownloadAPIClient(httpx.AsyncClient):
+    def __init__(
+        self,
+        progress=DEFAULT.progress,
+        raise_for_status=DEFAULT.raise_for_status,
+        **kwargs,
+    ):
+        defaults = dict(
+            base_url=API_SERVICE_ROOT,
+            event_hooks=dict(request=[], response=[]),
+            timeout=60,
+        )
+
+        if raise_for_status:
+            defaults["event_hooks"]["response"].append(_raise_for_status)
+        else:
+            defaults["event_hooks"]["response"].append(_warn_for_status)
+
+        defaults.update(kwargs)
+        super().__init__(**defaults)
+        self.progress = progress
+
+    async def raw_country_json(self):
+        """
+        Get list of country codes
+        """
+        response = await self.get("/Country")
+        return response.json()
+
+    async def raw_property_json(self):
+        """
+        Get list of pollutants and notation
+        """
+        response = await self.get("/Property")
+        return response.json()
+
+    async def raw_city_json(self, *countries: str):
+        response = await self.post("/City", json=countries)
+        return response.json()
+
+    async def raw_download_summary_json(self, url: DownloadInfo):
+        response = await self.post("/DownloadSummary", json=url.request_info())
+        return response.json()
+
+    async def total_num_parquet_files(self, *urls: DownloadInfo):
+        """
+        The total number of parquet files that will be downloaded for a set of URLs
+        """
+        total_files = 0
+        summary_jobs = [self.raw_download_summary_json(url) for url in urls]
+        for future in asyncio.as_completed(summary_jobs):
+            result: dict[str, int] = await future
+            total_files += result["numberFiles"]
+        return total_files
+
+    async def parquet_file_urls(self, *urls: DownloadInfo):
+        """
+        Get the Parquet file URLs matching all the provided DownloadInfo
+        """
+        jobs = [
+            self.post("/ParquetFile/urls", json=info.request_info())
+            for info in urls
+        ]
+
+        for response in tqdm(
+            asyncio.as_completed(jobs),
+            total=len(jobs),
+            leave=True,
+            disable=not self.progress or len(jobs) <= 1,
+            desc="Fetching Parquet file URLs",
+        ):
+            response = await response
+            lines = [line.strip() for line in response.text.split("\n")]
+            for line in lines:
+                if line.startswith("http"):
+                    yield line
+
+    async def pollutants(self) -> defaultdict[str, set[int]]:
```

Comment:
this could be a [@cached_property] as it is highly unlikely that the counties and pollutants change during a session.


[@cached_property]: https://docs.python.org/3/library/functools.html#functools.cached_property

Reply:
Yeah agreed, I cached the raw response but this makes more sense